src/test/app/PermissionedDomains_test.cpp

```diff
@@ -0,0 +1,504 @@
+        };
+        uint256 domain2;
+        {
+            BEAST_EXPECT(
+                credentials10.size() ==
+                maxPermissionedDomainCredentialsArraySize);
+            BEAST_EXPECT(
+                credentials10 != pdomain::sortCredentials(credentials10));
+            env(pdomain::setTx(alice[0], credentials10));
+            auto tx = env.tx()->getJson(JsonOptions::none);
+            domain2 = pdomain::getNewDomain(env.meta());
+            auto objects = pdomain::getObjects(alice[0], env);
+            auto object = objects[domain2];
+            BEAST_EXPECT(
+                pdomain::credentialsFromJson(object, pubKey2Acc) ==
+                pdomain::sortCredentials(credentials10));
+        }
+
+        // Update with 1 credential.
+        env(pdomain::setTx(alice[0], credentials1, domain2));
+        BEAST_EXPECT(
+            pdomain::credentialsFromJson(
+                pdomain::getObjects(alice[0], env)[domain2], pubKey2Acc) ==
+            credentials1);
+
+        // Update with 10 credentials.
+        env(pdomain::setTx(alice[0], credentials10, domain2));
+        env.close();
+        BEAST_EXPECT(
+            pdomain::credentialsFromJson(
+                pdomain::getObjects(alice[0], env)[domain2], pubKey2Acc) ==
+            pdomain::sortCredentials(credentials10));
+
+        // Update from the wrong owner.
+        env(pdomain::setTx(alice[2], credentials1, domain2),
+            ter(tecNO_PERMISSION));
+
+        // Update a uint256(0) domain
+        env(pdomain::setTx(alice[0], credentials1, uint256(0)),
+            ter(temMALFORMED));
+
+        // Update non-existent domain
+        env(pdomain::setTx(alice[0], credentials1, uint256(75)),
+            ter(tecNO_ENTRY));
+
+        // Test bad data when creating a domain.
+        testBadData(alice[0], env);
+        // Test bad data when updating a domain.
+        testBadData(alice[0], env, domain2);
+
+        // Try to delete the account with domains.
+        auto const acctDelFee(drops(env.current()->fees().increment));
+        constexpr std::size_t deleteDelta = 255;
+        {
+            // Close enough ledgers to make it potentially deletable if empty.
+            std::size_t ownerSeq = env.seq(alice[0]);
+            while (deleteDelta + ownerSeq > env.current()->seq())
+                env.close();
+            env(acctdelete(alice[0], alice[2]),
+                fee(acctDelFee),
+                ter(tecHAS_OBLIGATIONS));
+        }
+
+        {
+            // Delete the domains and then the owner account.
+            for (auto const& objs : pdomain::getObjects(alice[0], env))
+                env(pdomain::deleteTx(alice[0], objs.first));
+            env.close();
+            std::size_t ownerSeq = env.seq(alice[0]);
+            while (deleteDelta + ownerSeq > env.current()->seq())
+                env.close();
+            env(acctdelete(alice[0], alice[2]), fee(acctDelFee));
+        }
```

Comment:
Missing tests:
* `CredentialType` is an empty string or is too long
* `CredentialType` length boundary condition
* One account can create multiple PDs

Reply:
added